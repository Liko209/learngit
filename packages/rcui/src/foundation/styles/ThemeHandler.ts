import { EventEmitter2 } from 'eventemitter2';
import options from './options.json';

type Options = {
  loadPath: (theme: string) => string;
};

class ThemeHandler extends EventEmitter2 {
  private options: Options = {
    loadPath: theme => `${options.basePath}${theme}.json`,
  };

  async applyTheme(theme: string) {
    const themeOptions = await this.read(theme);
    this.emit(options.bindEvent, themeOptions);
  }

  read(theme: string) {
    const { loadPath } = this.options;
    const url = loadPath(theme);

    return this.loadUrl(url);
  }

  async loadUrl(url: string) {
    return fetch(url)
      .then((response: Response) => {
        if (response.ok) {
          try {
            return response.json();
          } catch (e) {
            return Promise.reject(`failed parsing ${url} to json`);
          }
        }
        return Promise.reject(`failed loading ${url}`);
      })
      .catch(() => Promise.reject(`failed loading ${url}`));
  }
}
export { ThemeHandler };
export default new ThemeHandler();
