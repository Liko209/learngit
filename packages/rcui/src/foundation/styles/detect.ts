import applyTheme from './ThemeHandler';
import options from './options.json';

async function detect() {
  let theme;
  try {
    theme = (await applyTheme.read(options.brandThemeFile)) as {
      [brand: string]: {
        default: string;
        themes: string[];
      };
    };
  } catch (error) {
    console.log(error);
  }
  return theme;
}

export default detect;
