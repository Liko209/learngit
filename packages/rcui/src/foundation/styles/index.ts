export { default as ThemeProvider } from './ThemeProvider';
export { default as detect } from './detect';
export { default as themeHandler } from './ThemeHandler';
export * from './theme.d';
