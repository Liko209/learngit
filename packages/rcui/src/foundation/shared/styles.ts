import { Theme } from '../styles';
import { css, keyframes } from '../styled-components';

function ellipsis() {
  return css`
    white-space: nowrap;
    overflow: hidden;
    text-overflow: ellipsis;
    word-break: break-word;
  `;
}

function lineClamp(lineNumber: number, maxHeight: number) {
  return css`
    overflow: hidden;
    text-overflow: ellipsis;
    display: -webkit-box;
    -webkit-line-clamp: ${lineNumber};
    -webkit-box-orient: vertical;
    max-height: ${maxHeight};
    word-break: break-word;
  `;
}

const rippleEnter = (theme: Theme) => keyframes`
  from {
    transform: scale(0);
    opacity: 0.1;
  }
  to {
    transform: scale(1);
    opacity: ${theme.opacity[5]};
  }
`;

export { ellipsis, lineClamp, rippleEnter };
