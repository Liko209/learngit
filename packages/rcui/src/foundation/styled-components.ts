import * as styledComponents from 'styled-components';
import {
  ThemedStyledComponentsModule,
  ThemeProps as StyledThemeProps,
} from "styled-components"; // tslint:disable-line

import { Theme } from './styles';

const {
  default: styled,
  css,
  keyframes,
  createGlobalStyle,
  withTheme,
  ThemeProvider,
  ThemeConsumer,
} = styledComponents as ThemedStyledComponentsModule<Theme>;

type ThemeProps = StyledThemeProps<Theme>;

export {
  css,
  createGlobalStyle,
  keyframes,
  ThemeProvider,
  withTheme,
  ThemeConsumer,
  ThemeProps,
};

export default styled;
