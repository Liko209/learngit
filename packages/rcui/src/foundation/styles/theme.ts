import palette from './theme/palette.json';
import typography from './theme/typography.json';
import opacity from './theme/opacity.json';
import radius from './theme/radius.json';
import spacing from './theme/spacing.json';
import zIndex from './theme/zIndex.json';
import breakpoints from './theme/breakpoints.json';
import boxShadow from './theme/boxShadow.json';

import {
  Theme as MuiTheme,
  ThemeOptions,
} from '@material-ui/core/styles/createMuiTheme';
import { Transitions } from '@material-ui/core/styles/transitions';
import { Shadows } from '@material-ui/core/styles/shadows';
import { ZIndex as MuiZIndex } from '@material-ui/core/styles/zIndex';
import { Spacing as MuiSpacing } from '@material-ui/core/styles/spacing';
import { Typography as MuiTypography } from '@material-ui/core/styles/createTypography';
import {
  Palette as MuiPalette,
  PaletteColor,
} from '@material-ui/core/styles/createPalette';

export type Palette = { [P in keyof typeof palette]: PaletteColor } &
  MuiPalette;
export type Typography = typeof typography & MuiTypography;
export type Opacity = typeof opacity;
export type Radius = typeof radius;
export type Spacing = typeof spacing & MuiSpacing;
export type ZIndex = typeof zIndex & MuiZIndex;
export type boxShadow = typeof boxShadow;
export type size = {
  width: number;
  height: number;
};

export { Transitions, Shadows };

export type Theme = {
  palette: Palette;
  typography: Typography;
  opacity: Opacity;
  radius: Radius;
  spacing: Spacing;
  zIndex: ZIndex;
  tooltip: {
    dark: string;
  };
  size: size;
  boxShadow: boxShadow;
  shape: {
    border1: string;
    border2: string;
    border3: string;
    border4: string;
  };
} & MuiTheme;

export type PaletteOptions = typeof palette;
export type TypographyOptions = typeof typography;
export type OpacityOptions = Opacity;
export type RadiusOptions = Radius;
export type SpacingOptions = typeof spacing;
export type ZIndexOptions = typeof zIndex;
export type BreakpointsOptions = typeof breakpoints;

export type ThemeOptions = {
  palette: PaletteOptions;
  typography: TypographyOptions;
  opacity: OpacityOptions;
  radius: OpacityOptions;
  spacing: SpacingOptions;
  zIndex: ZIndexOptions;
  breakpoints: BreakpointsOptions;
};
