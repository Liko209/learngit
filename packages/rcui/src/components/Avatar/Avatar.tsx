/*
 * @Author: Shining Miao (shining.miao@ringcentral.com)
 * @Date: 2018-08-17 10:25:50
 * Copyright Â© RingCentral. All rights reserved.
 */
import React, { PureComponent } from 'react';
import styled from '../../foundation/styled-components';
import MuiAvatar, {
  AvatarProps as MuiAvatarProps,
} from '@material-ui/core/Avatar';
import { typography, palette, opacity } from '../../foundation/shared/theme';
import dataColors from '../../foundation/shared/colors.json';
import { Theme } from '../../foundation/styles';

type Size = 'xs' | 'sm' | 'md' | 'lg';

type AvatarProps = {
  size: Size;
  color?: string;
  presence?: JSX.Element;
};

type RuiAvatarProps = AvatarProps & MuiAvatarProps;

const avatarSize = {
  xs: '24px',
  sm: '32px',
  md: '40px',
  lg: '80px',
};

const fonts: { [key in Size]: keyof Theme['typography'] } = {
  lg: 'display1',
  md: 'subheading2',
  sm: 'subheading2',
  xs: 'caption2',
};

const StyledWrapper = styled.div<{ size: Size }>`
  width: ${({ size }) => avatarSize[size]};
  height: ${({ size }) => avatarSize[size]};
  position: relative;
`;

const StyledAvatar = styled(MuiAvatar)<RuiAvatarProps>`
  && {
    background-color: ${({ color }) =>
      color ? dataColors[color] : palette('common', 'white')};
    width: ${({ size }) => avatarSize[size]};
    height: ${({ size }) => avatarSize[size]};
    ${({ size }) => typography(fonts[size])};
    &:hover {
      opacity: ${opacity('8')};
      cursor: pointer;
    }
    &:active {
      opacity: ${opacity('6')};
    }
  }
` as typeof MuiAvatar;

const StyledPresenceWrapper = styled.div`
  position: absolute;
  bottom: 0;
  right: 0;
`;

class RuiAvatar extends PureComponent<RuiAvatarProps> {
  static defaultProps = {
    size: 'md',
  };

  render() {
    const { presence, size } = this.props;

    return presence ? (
      <StyledWrapper size={size}>
        <StyledAvatar {...this.props} />
        <StyledPresenceWrapper>{presence}</StyledPresenceWrapper>
      </StyledWrapper>
    ) : (
      <StyledAvatar {...this.props} />
    );
  }
}

export { RuiAvatarProps, RuiAvatar };
