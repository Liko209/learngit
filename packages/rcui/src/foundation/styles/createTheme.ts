import deepmerge from 'deepmerge';
import palette from './theme/palette.json';
import typography from './theme/typography.json';
import opacity from './theme/opacity.json';
import radius from './theme/radius.json';
import spacing from './theme/spacing.json';
import zIndex from './theme/zIndex.json';
import breakpoints from './theme/breakpoints.json';
import { createMuiTheme } from '@material-ui/core/styles';
import createMuiPalette from '@material-ui/core/styles/createPalette';
import { Theme, ThemeOptions, PaletteOptions } from './theme';

function createPalette(paletteOptions: PaletteOptions) {
  const { augmentColor } = createMuiPalette(paletteOptions);
  Object.keys(paletteOptions).forEach((key: keyof PaletteOptions) => {
    if (palette[key]) {
      augmentColor(paletteOptions[key]);
    }
  });

  return paletteOptions;
}

function createTheme(options: Partial<ThemeOptions> = {}) {
  const {
    palette: paletteInput = {},
    typography: typographyInput = {},
    opacity: opacityInput = {},
    radius: radiusInput = {},
    spacing: spacingInput = {},
    zIndex: zIndexInput = {},
    breakpoints: breakpointsInput = {},
  } = options;

  // suppress material-ui warning
  // @ts-ignore
  typographyInput.suppressWarning = true;

  const theme = deepmerge(
    {
      palette: createPalette(deepmerge(palette, paletteInput)),
      typography: deepmerge(typography, typographyInput),
      opacity: deepmerge(opacity, opacityInput),
      radius: deepmerge(radius, radiusInput),
      spacing: deepmerge(spacing, spacingInput),
      zIndex: deepmerge(zIndex, zIndexInput),
      breakpoints: deepmerge(breakpoints, breakpointsInput),
    },
    options,
  );

  return createMuiTheme(theme) as Theme;
}

export default createTheme;
