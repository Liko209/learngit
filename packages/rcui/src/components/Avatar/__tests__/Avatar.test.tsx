import * as React from 'react';
import * as renderer from 'react-test-renderer';
import { RuiAvatar } from '../Avatar';
import { ThemeProvider } from '../../../foundation/styles';

describe('Avatar', () => {
  it('renders correctly', () => {
    const tree = renderer
      .create(
        <ThemeProvider>
          <RuiAvatar color="lake">SH</RuiAvatar>
        </ThemeProvider>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('mount correctly', () => {
    expect(() =>
      renderer.create(
        <ThemeProvider>
          <RuiAvatar color="lake">SH</RuiAvatar>
        </ThemeProvider>,
      ),
    ).not.toThrow();
  });
});
