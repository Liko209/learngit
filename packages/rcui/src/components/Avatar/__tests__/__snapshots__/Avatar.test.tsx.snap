// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Avatar renders correctly 1`] = `
.c0.c0 {
  background-color: #69A3EB;
  width: 40px;
  height: 40px;
  font-size: 1rem;
  font-weight: 500;
  line-height: 22px;
}

.c0.c0:hover {
  opacity: 0.8;
  cursor: pointer;
}

.c0.c0:active {
  opacity: 0.6;
}

<div
  className="MuiAvatar-root-1 MuiAvatar-colorDefault-2 c0"
  color="lake"
  size="md"
>
  SH
</div>
`;
